File: src/main/java/org/taktik/icure/services/external/rest/v1/dto/filter/contact/ContactByServiceIdsFilter.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.services.external.rest.v1.dto.filter.contact;

import java.util.List;

import com.google.common.base.Objects;
import org.taktik.icure.entities.Contact;
import org.taktik.icure.services.external.rest.handlers.JsonPolymorphismRoot;
import org.taktik.icure.services.external.rest.v1.dto.filter.Filter;

@JsonPolymorphismRoot(Filter.class)
public class ContactByServiceIdsFilter extends Filter<Contact> implements org.taktik.icure.dto.filter.contact.ContactByServiceIdsFilter {
    List<String> ids;

    public ContactByServiceIdsFilter() {
    }

    public ContactByServiceIdsFilter(List<String> ids) {
        this.ids = ids;
    }

    @Override
    public List<String> getIds() {
        return ids;
    }

    public void setIds(List<String> ids) {
        this.ids = ids;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        ContactByServiceIdsFilter that = (ContactByServiceIdsFilter) o;

        return Objects.equal(this.ids, that.ids);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(ids);
    }

	@Override
	public boolean matches(Contact item) {
		return item.getServices().stream().filter(s->ids.contains(s.getId())).findAny().isPresent();
	}
}

File: src/main/kotlin/org/taktik/icure/domain/filter/impl/contact/ContactByServiceIdsFilter.kt
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.taktik.icure.domain.filter.impl.contact

import com.github.pozo.KotlinBuilder
import com.google.common.base.Objects
import org.taktik.icure.domain.filter.AbstractFilter
import org.taktik.icure.entities.Contact

@KotlinBuilder
data class ContactByServiceIdsFilter(
        override val desc: String? = null,
        override val ids: List<String>? = null
) : AbstractFilter<Contact>, org.taktik.icure.domain.filter.contact.ContactByServiceIdsFilter {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val filter = other as ContactByServiceIdsFilter
        return Objects.equal(ids, filter.ids)
    }

    override fun hashCode(): Int {
        return Objects.hashCode(ids)
    }

    override fun matches(item: Contact): Boolean {
        return item.services.stream().filter { (id) -> ids!!.contains(id) }.findAny().isPresent
    }
}
